compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -373,11 +413,108 @@ impl<'tcx> SizeSkeleton<'tcx> {
             (SizeSkeleton::Pointer { tail: a, .. }, SizeSkeleton::Pointer { tail: b, .. }) => {
                 a == b
             }
+            // constants are always pre-normalized into a canonical form so this
+            // only needs to check if their pointers are identical.
+            (SizeSkeleton::Generic(a), SizeSkeleton::Generic(b)) => a == b,
             _ => false,
         }
     }
 }
 
+/// To compare constants, we only use this to multiply them, which ensures
```

Comment:
This needs a more elaborate comment I feel given that this code isn't so straight forward. Can you maybe give an example and mention the intent behind this function?